File: src/host/go/rvm.go
package main

import (
	"fmt"
	"io" // @@(feature go/io)@@
	"os" // @@(feature go/os)@@
	"runtime" // @@(feature go/runtime)@@
	"syscall" // @@(feature go/syscall)@@
)

const DebugICall = false

//goland:noinspection SpellCheckingInspection
// @@(replace ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" (encode 92)
const Input = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" // RVM code that prints HELLO!
// )@@

func ShowRib(rib Obj, depth int) {
  if depth < 0 {
    if rib.Number() {
      fmt.Printf("...")
    } else {
      fmt.Printf("[...]")
    }
    return
  }

  if rib.Number() {
    fmt.Printf("%d", rib.Value())
    return
  }
  if (rib == TRUE) {
    fmt.Printf("#t")
    return
  }
  if (rib == FALSE) {
    fmt.Printf("#f")
    return
  }
  if (rib == NIL) {
    fmt.Printf("'()")
    return
  }
  fmt.Printf("[")
  ShowRib(rib.Field0(), depth - 1)
  fmt.Printf(", ")
  ShowRib(rib.Field1(), depth - 1)
  fmt.Printf(", ")
  ShowRib(rib.Field2(), depth - 1)
  fmt.Printf("]")
}

const (
	InstrCall  = 0
	InstrSet   = 1
	InstrGet   = 2
	InstrConst = 3
	InstrIf    = 4
	InstrHalt  = 5
)

const (
	PairTag      = 0
	ClosureTag   = 1
	SymbolTag    = 2
	StringTag    = 3
	SingletonTag = 5
)

// ===============================================
// ===============================================
//                 OBJ + RIB
//                   + prim
// ===============================================
// ===============================================

type prim1 func(x Obj) Obj
type prim2 func(x, y Obj) Obj
type prim3 func(x, y, z Obj) Obj

type Obj interface {
	Number() bool
	Rib() bool
	Field0() Obj
	Field0Set(Obj) Obj
	Field1() Obj
	Field1Set(Obj) Obj
	Field2() Obj
	Field2Set(Obj) Obj
	Value() int
	Add(int) Obj
}

type Num struct {
	x int
}

type Rib struct {
	x, y, z Obj
}

func tagNum(x int) Obj {
	nb := new(Num)
	nb.x = x
	return nb
}

func (num *Num) Number() bool {
	return true
}

func (num *Num) Rib() bool {
	return false
}

func (num *Num) Field0() Obj {
	panic("Cannot call field0 on non-primitive")
}

func (num *Num) Field1() Obj {
	panic("Cannot call field1 on non-primitive")
}

func (num *Num) Field2() Obj {
	panic("Cannot call field2 on non-primitive")
}

func (num *Num) Field0Set(Obj) Obj {
	panic("Cannot call field0set on non-primitive")
}

func (num *Num) Field1Set(Obj) Obj {
	panic("Cannot call field1set on non-primitive")
}

func (num *Num) Field2Set(Obj) Obj {
	panic("Cannot call field2set on non-primitive")
}

func (num *Num) Value() int {
	return num.x
}

func (num *Num) Add(v int) Obj {
	r := num.x + v
	return tagNum(r)
}

func (*Rib) Number() bool {
	return false
}

func (*Rib) Rib() bool {
	return true
}

func (trio *Rib) Field0() Obj {
	return trio.x
}

func (trio *Rib) Field1() Obj {
	return trio.y
}

func (trio *Rib) Field2() Obj {
	return trio.z
}

func (trio *Rib) Field0Set(obj Obj) Obj {
	trio.x = obj
	return obj
}

func (trio *Rib) Field1Set(obj Obj) Obj {
	trio.y = obj
	return obj
}

func (trio *Rib) Field2Set(obj Obj) Obj {
	trio.z = obj
	return obj
}

func (*Rib) Value() int {
	panic("Not a number")
}

func (*Rib) Add(int) Obj {
	panic("Not a number")
}

// ===============================================
// ===============================================
//                Input PROCESSING
// ===============================================
// ===============================================

var pos = 0

func getByte() byte {
	r := Input[pos]
	pos++
	return r
}

func getCode() int {
	x := int(getByte()) - 35

	if x < 0 {
		return 57
	} else {
		return x
	}
}

func getInt(n int) int {
	x := getCode()
	n *= 46

	if x < 46 {
		return n + x
	} else {
		return getInt(n + x - 46)
	}
}

// ===============================================
// ===============================================
//                    CONSTANTS
// ===============================================
// ===============================================

var NUM0 = tagNum(0)
var stack = NUM0
var symbolTable = NUM0
var pc = NUM0

var FALSE *Rib = nil
var TRUE *Rib = nil
var NIL *Rib = nil

func push(val Obj) {
	tos := new(Rib)
	tos.x = val
	tos.y = stack
	tos.z = tagNum(PairTag)

	stack = tos
}

func pop() Obj {
	x := stack.Field0()
	stack = stack.Field1()
	return x
}

func allocRib(car, cdr, tag Obj) Obj {
	push(car)
	allocated := stack

	oldStack := allocated.Field1()
	stack = oldStack

	allocated.Field1Set(cdr)
	allocated.Field2Set(tag)

	return allocated
}

func lstLength(lst Obj) Obj {
	n := 0

	for lst.Rib() && lst.Field2().Value() == 0 {
		n++
		lst = lst.Field1()
	}

	return tagNum(n)
}

func createSym(chars Obj) Obj {
	str := allocRib(chars, lstLength(chars), tagNum(StringTag))
	sym := allocRib(FALSE, str, tagNum(SymbolTag))
	return allocRib(sym, symbolTable, tagNum(PairTag))
}

func listTail(list, i Obj) Obj {
	if i.Value() == 0 {
		return list
	} else {
		return listTail(list.Field1(), i.Add(-1))
	}
}

func instTail(list, i Obj) Obj {
	if i.Value() == 0 {
		return list
	} else {
		return listTail(list.Field2(), i.Add(-1))
	}
}

func listRef(list, i Obj) Obj {
	return listTail(list, i).Field0()
}

func symbolRef(depth Obj) Obj {
	return listRef(symbolTable, depth)
}

func setGlobal(c Obj) {
	symbolTable.Field0().Field0Set(c)
	symbolTable = symbolTable.Field1()
}

func getOperand(o Obj) Obj {
	var op Obj
	if o.Number() {
		op = listTail(stack, o)
	} else {
		op = o
	}

	return op.Field0()
}

func buildSymTable() {
	n := getInt(0)

	for n > 0 {
		n--
		symbolTable = createSym(NIL)
	}

	accum := Obj(NIL)

	for {
		c := int(getByte())

		if c == 44 {
			symbolTable = createSym(accum)
			accum = NIL
			continue
		}

		if c == 59 {
			break
		}

		accum = allocRib(tagNum(c), accum, tagNum(PairTag))
	}

	symbolTable = createSym(accum)
}

// @@(feature encoding/optimal
func decode() {
	var ranges = []int{1, 2, 3} // @@(replace "{1, 2, 3}" (list->host encoding/optimal/start "{" "," "}"))@@

	//var n Obj
	var i int
	var arg Obj

	for {
		code := getCode()
		arg = tagNum(code)
		range_index := 0

		for {
			if arg.Value() < ranges[range_index] {
				break
			}

			arg = tagNum(arg.Value() - ranges[range_index])
			range_index++
		}

    if (range_index < 4) { push(tagNum(0)) } // JUMP
    if (range_index < 24) {
			if range_index%2>0{
				arg = tagNum(getInt(arg.Value()))
			}
		}

		if range_index < 20 { // jump call set get const
			i = (range_index / 4) - 1
			if i < 0 {
				i = 0
			}
			if (range_index % 4) / 2 >= 1{
				arg = symbolRef(arg)
			} else {
				arg = arg
			}
		} else if range_index < 22 { // const proc
			arg = allocRib(allocRib(arg, NUM0, pop()), Obj(NIL), tagNum(ClosureTag))
			i=3;
			if stack == NUM0 {
				break
			}
		} else if range_index < 24 { // skip
			stack = allocRib(instTail(stack.Field0(), arg), stack, NUM0)
			continue
		} else if (range_index < 25) { // if
			arg = pop()
			i=4;
		}

		stack.Field0Set(allocRib(tagNum(i), arg, stack.Field0()))
	}

	pc = arg.Field0().Field2()
}
// )@@


// @@(feature encoding/original
func decode() {
	weights := []int{20, 30, 0, 10, 11, 4}
	var n Obj
	var d, op int

	for {
		x := getCode()
		n = tagNum(x)
		op = -1

		for {
			op++
			d = weights[op]

			if n.Value() > (2 + d) {
				n = n.Add(-(d + 3))
			} else {
				break
			}
		}

		if x > 90 {
			op = InstrIf
			n = pop()
		} else {
			if op == 0 {
				stack = allocRib(tagNum(0), stack, tagNum(PairTag))
			}

			if n.Value() >= d {
				if n.Value() == d {
					n = tagNum(getInt(0))
				} else {
					n = symbolRef(tagNum(getInt(n.Value() - d - 1)))
				}
			} else {
				if op < 3 {
					n = symbolRef(n)
				}
			}

			if op > 4 {
				inner := allocRib(n, tagNum(0), pop())
				n = allocRib(inner, NIL, tagNum(ClosureTag))

				if stack == nil || (stack.Number() && stack.Value() == 0) {
					break
				}
				op = InstrConst
			} else if op > 0 {
				op--
			} else {
				op = 0
			}
		}

		stack.Field0Set(allocRib(tagNum(op), n, stack.Field0()))
	}

	pc = n.Field0().Field2()
}
// )@@

func initConstantRibs() {

	init := func(x, y, z int) *Rib {
		cl := new(Rib)
		cl.x = tagNum(x)
		cl.y = tagNum(y)
		cl.z = tagNum(z)
		return cl
	}

	TRUE = init(0, 0, SingletonTag)
	FALSE = init(0, 0, SingletonTag)
	NIL = init(0, 0, SingletonTag)
}

func setupStack() {
	stack = allocRib(
		tagNum(0),
		tagNum(0),
		allocRib(tagNum(InstrHalt),
			tagNum(0),
			tagNum(PairTag)))

}

func doPrim1(f prim1) {
	x := pop()
	push(f(x))
}

func doPrim2(f prim2) {
	y := pop()
	x := pop()
	push(f(x, y))
}

func doPrim3(f prim3) {
	z := pop()
	y := pop()
	x := pop()
	push(f(x, y, z))
}

func boolean(x bool) Obj {
	if x {
		return TRUE
	} else {
		return FALSE
	}
}

// @@(feature scm2str
func chars2str(chars Obj) string {
	if chars == NIL {
		return ""
	} else {
		return string(byte(chars.Field0().Value())) + chars2str(chars.Field1())
	}
}

func scm2str(s Obj) string {
	return chars2str(s.Field0())
}
// )@@

// @@(feature str2scm
func str2scm(c string) Obj {
	obj := Obj(NIL)
	length := len(c)

	for i := length - 1; i >= 0; i-- {
		obj = allocRib(tagNum(int(c[i])), obj, tagNum(PairTag))
	}

	return allocRib(obj, tagNum(length), tagNum(StringTag))
}
// )@@

// @@(feature list2scm (use str2scm)
func list2scm(s []string) Obj {
	obj := Obj(NIL)
	for i := len(s) - 1; i >= 0; i-- {
		obj = allocRib(str2scm(s[i]), obj, tagNum(PairTag))
	}

	return obj
}
// )@@


func prim(primNo int) Obj {

	if DebugICall {
		fmt.Printf("Calling primitive %d\n", primNo)
	}

	switch primNo {
		// @@(primitives (gen "case " index ":" body)
	case 0: // @@(primitive (%%rib a b c)
		doPrim3(func(x, y, z Obj) Obj {
			return allocRib(x, y, z)
		}) // )@@
	case 1: // @@(primitive (%%id x)
		doPrim1(func(x Obj) Obj {
			return x
		}) // )@@
	case 2: // @@(primitive (%%arg1 x y)
		pop() // )@@
	case 3: // @@(primitive (%%arg2 x y)
		x := pop()
		pop()
		push(x) // )@@
	case 4: // @@(primitive (%%close rib)
		x := stack.Field0().Field0()
		y := stack.Field1()
		z := tagNum(ClosureTag)
		stack.Field0Set(allocRib(x, y, z)) // )@@
	case 5: // @@(primitive (%%rib? rib)
		doPrim1(func(x Obj) Obj {
			return boolean(x.Rib())
		}) // )@@
	case 6: // @@(primitive (%%field0 rib)
		doPrim1(func(x Obj) Obj {
			return x.Field0()
		}) // )@@
	case 7: // @@(primitive (%%field1 rib)
		doPrim1(func(x Obj) Obj {
			return x.Field1()
		}) // )@@
	case 8: // @@(primitive (%%field2 rib)
		doPrim1(func(x Obj) Obj {
			return x.Field2()
		}) // )@@
	case 9: // @@(primitive (%%field0-set! rib val)
		doPrim2(func(x, y Obj) Obj {
			return x.Field0Set(y)
		}) // )@@
	case 10: // @@(primitive (%%field1-set! rib val)
		doPrim2(func(x, y Obj) Obj {
			return x.Field1Set(y)
		}) // )@@
	case 11: // @@(primitive (%%field2-set! rib val)
		doPrim2(func(x, y Obj) Obj {
			return x.Field2Set(y)
		}) // )@@
	case 12: // @@(primitive (%%eqv? x y)
		doPrim2(func(x, y Obj) Obj {
			if x.Number() && y.Number() {
				return boolean(x.Value() == (y.Value()))
			} else if x.Rib() && y.Rib() {
				return boolean(x == y)
			} else {
				return FALSE
			}
		}) // )@@
	case 13: // @@(primitive (%%< x y)
		doPrim2(func(x, y Obj) Obj {
			return boolean(x.Value() < y.Value())
		}) // )@@
	case 14: // @@(primitive (%%+ x y)
		doPrim2(func(x, y Obj) Obj {
			return tagNum(x.Value() + y.Value())
		}) // )@@
	case 15: // @@(primitive (%%- x y)
		doPrim2(func(x, y Obj) Obj {
			return tagNum(x.Value() - y.Value())
		}) // )@@
	case 16: // @@(primitive (%%* x y)
		doPrim2(func(x, y Obj) Obj {
			return tagNum(x.Value() * y.Value())
		}) // )@@
	case 17: // @@(primitive (%%quotient x y)
		doPrim2(func(x, y Obj) Obj {
			return tagNum(x.Value() / y.Value())
		}) // )@@
	case 18: // @@(primitive (%%getchar) (use go/io go/os)
		if pos < len(Input) {
			push(tagNum(int(getByte())))
		} else {
			buff := make([]byte, 1)
			count, err := os.Stdin.Read(buff)


      if err == io.EOF {
			  push(tagNum(int(-1)))
      } else if nil != err {
				panic(err)
			} else if count != 1 {
				panic("Failed to read 1char")
			} else {
			  push(tagNum(int(buff[0])))
      }

		} // )@@
	case 19: // @@(primitive (%%putchar x)
		doPrim1(func(x Obj) Obj {
			fmt.Printf("%c", byte(x.Value()))
			return x
		}) //)@@
// )@@
	}

	return tagNum(0)
}

func getCont() Obj {
	s := stack

	for s.Field2().Number() && s.Field2().Value() == 0 {
		s = s.Field1()
	}

	return s
}

func run() {
	for {
		instr := pc.Field0().Value()
		operand := pc.Field1()
		next := pc.Field2()

		switch instr {
		case InstrCall: // call
			call := next.Rib()
			if DebugICall {
				if call {
					fmt.Println("--- call")
				} else {
					fmt.Println("--- jump")
				}
			}

			proc := getOperand(operand)

			for {
			  code := proc.Field0()

			  if code.Number() {
			  	pop(); // @@(feature arity-check)@@

			  	ret := prim(code.Value())

					if ret.Rib() {
						proc = ret
						continue
					}

			  	if call {
			  		code = pc
			  	} else {
			  		code = getCont()
			  		stack.Field1Set(code.Field0())
			  	}
			  } else {
			  	if DebugICall {
			  		fmt.Printf("Calling a symbol\n")
			  	}

			  	nargs := pop().Value() // @@(feature arity-check)@@

			  	c2 := allocRib(tagNum(0), proc, tagNum(PairTag))
			  	s2 := c2

			  	arityNumber := code.Field0().Value()
			  	nparams := arityNumber >> 1

			  	// @@(feature arity-check
			  	{
			  		var shouldCrash bool
			  		if arityNumber & 1 == 1 {
			  			shouldCrash = nparams > nargs
			  		} else {
			  			shouldCrash = nparams != nargs
}
			  		if shouldCrash {
			  			panic(fmt.Sprintf("Arity mismatch: expected %d, got %d, arityNumber=%d", nparams, nargs, arityNumber))
			  		}
			  	}
			  	// )@@

			  	// @@(feature rest-param (use arity-check)
			  	nargs -= nparams
			  	if arityNumber & 1 == 1 {
			  		rest := Obj(NIL)
			  		for nargs > 0 {
			  			rest = allocRib(pop(), rest, tagNum(PairTag))
			  			nargs--
			  		}

			  		s2 = allocRib(rest, s2, tagNum(PairTag))
			  	}
			  	// )@@

			  	for nparams > 0 {
			  		nparams--
			  		s2 = allocRib(pop(), s2, tagNum(PairTag))
			  	}

			  	if call {
			  		c2.Field0Set(stack)
			  		c2.Field2Set(pc.Field2())
			  	} else {
			  		cont := getCont()
			  		c2.Field0Set(cont.Field0())
			  		c2.Field2Set(cont.Field2())
			  	}

			  	stack = s2
			  }

			  pc = code.Field2()
			  break
		  }
		case InstrSet: // set
			if DebugICall {
				fmt.Println("--- set")
			}
	    x := stack.Field0()

			if pc.Field1().Number() {
				listTail(stack, pc.Field1()).Field0Set(x)
			} else {
				pc.Field1().Field0Set(x)
			}
	    stack = stack.Field1()
			pc = pc.Field2()
		case InstrGet: // get
			if DebugICall {
				fmt.Println("--- get")
			}
			push(getOperand(pc.Field1()))
			pc = pc.Field2()
		case InstrConst: // const
			if DebugICall {
				fmt.Println("--- const")
			}
			push(pc.Field1())
			pc = pc.Field2()
		case InstrIf: // if
			if DebugICall {
				fmt.Println("--- if")
			}
			p := pop()

			if p == FALSE {
				pc = pc.Field2()
			} else {
				pc = pc.Field1()
			}
		default:
			fmt.Printf("Unknown instruction: %d\n", instr)
			fallthrough
		case InstrHalt:
			return
		}
	}
}

func initVm() {
	initConstantRibs()
	buildSymTable()
	decode()

	setGlobal(allocRib(tagNum(0), symbolTable, tagNum(ClosureTag)))
	setGlobal(FALSE)
	setGlobal(TRUE)
	setGlobal(NIL)

	setupStack()

	run()
}

func main() {
	initVm()
}

File: src/host/java/rvm.java
import java.io.IOException;
import java.util.Arrays;

class RVM {
    public static int pos = -1;
    static Object stack = new Object[]{0, 0,0};
    static Object[] FALSE = new Object[]{0,0,5,0};
    static Object[] TRUE = new Object[]{0,0,5,1};
    static Object[] NIL = new Object[]{0,0,5};
    static Object x;
    static Object y;
    static Object z;
    static Object[] symtbl;
    public static void main(String[] args) {


        // @@(replace ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" (encode 92)
        String input = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y";
        // )@@
        
        symtbl = NIL;
        int n = get_int(0,input);
        while (n > 0){
            n-=1;
            symtbl = new Object[]{new Object[]{FALSE,new Object[]{NIL,0,3},2},symtbl,0};
        }
        Object accum = NIL;
        n = 0;
        Object nArray = null;
        while (true){
            int c = get_byte(input);
            if (c == 44){
                symtbl = new Object[]{new Object[]{FALSE,new Object[]{accum,n,3},2},symtbl,0};accum=NIL;n=0;
            }else {
                if (c==59){
                    Object[]symbol_to_append = new Object[]{FALSE,new Object[]{accum,n,3},2};
                    symtbl = new Object[]{symbol_to_append,symtbl,0};
                    break;
                }
                accum = new Object[]{c,accum,0};
                n++;
            }
        }
        int[] temp_num = {20,30,0,10,11,4};
        while(true){
            int x = get_Code(input);
            n = x;
            int d = 0;
            int op = 0;
            while (true){
                d = temp_num[op];
                if (n <=2+d)break;
                n-=d+3;op+=1;
            }
            if (x > 90) nArray = pop();
            else{
                if (op ==0) {
                    stack = new Object[]{0,stack,0};
                    op+=1;
                }
                if(n>=d){
                    if(n==d){
                        nArray = get_int(0,input);
                    }
                    else{
                        nArray = ((Object[]) list_tail(get_int(n-d-1,input),symtbl))[0];
                    }
                }
                else {
                    if (op < 3){
                        nArray = ((Object[]) list_tail(n,symtbl))[0];
                    }
                    else{
                        nArray = n;
                    }
                }
                if (4 < op){
                    nArray = new Object[]{new Object[]{nArray,0,pop()},NIL,1};                                   // Array
                    if (stack==NIL || ((Object[])stack)[1]==(Integer)0)break;
                    op = 4;
                }
            }
            ((Object[]) stack)[0] = new Object[]{op - 1, nArray, ((Object[]) stack)[0]};
        }
        Object[] temp = (Object[]) ((Object[]) nArray)[0];
        Object[] pc = (Object[]) temp[2];
        set_global(new Object[]{0, symtbl, 1});
        set_global(FALSE);
        set_global(TRUE);
        set_global(NIL);

        stack = new Object[]{0, 0, new Object[]{5, 0, 0}};
        int j = 0;
        while (true){
            Object o = pc[1];
            Object i = pc[0];
            if  ((int) i < 1){
                Object temp2 = ((Object[]) get_opnd(o))[0];
                while (true){
                    Object temp3;
                    if (temp2 instanceof Object[]){
                        temp3 = ((Object[])temp2)[0];
                    }
                    else{
                        temp3 = temp2;
                    }
                    if (is_rib(temp3)){
                        Object[] c = (Object[]) temp3;
                        Object[] c2 = new Object[]{0, temp2, 0};
                        Object[] s2 = c2;
                        int nparams = ((int) c[0]) >> 1;
                        while (nparams > 0){
                            s2 = new Object[]{pop(), s2, 0};
                            nparams -= 1;
                        }
                        if (pc[2] instanceof Object[]){
                            c2[0] = stack;
                            c2[2] = pc[2];
                        }
                        else {
                            Object[] k = get_cont();
                            c2[0] = k[0];
                            c2[2] = k[2];
                        }
                        stack = s2;
                        pc = c;
                        break;
                    }
                    else {
                        Object[] c;
                        int cInt = (int)temp3;
                        prim(cInt);
                        if (pc[2] instanceof Object[]){
                            c = pc;
                        }
                        else{
                            c = get_cont();
                            ((Object[])stack)[1] = c[0];
                        }
                        pc=c;
                        break;
                    }

                }
            }
            else if((int)i < 2){
                ((Object[])get_opnd(o))[0] = ((Object[])stack)[0];
                stack = ((Object[])stack)[1];
            }
            else if ((int)i < 3){
                push(((Object[])get_opnd(o))[0]);
            }
            else if ((int)i < 4){
                push(o);
            }
            else if ((int)i < 5){
                Object tempPop = pop();
                if(!(tempPop==FALSE)){
                    pc = (Object[]) pc[1];
                    continue;
                }
            }
            else{
                break;
            }
            pc = (Object[]) pc[2];
        }
    }
    private static Object get_opnd(Object o){
        return is_rib(o) ? o :  list_tail((int) o, ((Object[])stack));
    }

    public static Object[] get_cont(){
        Object[] s = ((Object[])stack);
        while (!(s[2] instanceof Object[])){
            s = (Object[]) s[1];
        }
        return s;
    }
    public static void set_global(Object val){
        Object[] temp = (Object[]) symtbl[0];
        temp[0] = val;
        symtbl =  (Object[]) symtbl[1];
    }
    private static void prim1(){
        x = pop();
    }
    private static void prim2(){
        y = pop();
        prim1();
    }
    private static void prim3(){
        z = pop();
        prim2();
    }
    public static void prim(int no){
        switch (no){
            case 0 -> {prim3();push(new Object[]{x,y,z});break;}
            case 1 -> {prim1();push(x);break;}
            case 2 -> {pop();break;}
            case 3 -> {x = pop();pop();push(x);break;}
            case 4 -> {Object[] temp = (Object[]) pop();push(new Object[]{temp[0],stack,1});break;}
            case 5 -> {prim1();push(bool2scm(is_rib(x)));break;}
            case 6 -> {prim1();push(((Object[]) x)[0]);break;}
            case 7 -> {prim1();push(((Object[]) x)[1]);break;}
            case 8 -> {prim1();push(((Object[]) x)[2]);break;}
            case 9 -> {prim2();push(f0s(y,(Object[]) x));break;}
            case 10 -> {prim2();push(f1s(y,(Object[]) x));break;}
            case 11 -> {prim2();push(f2s(y,(Object[]) x));break;}
            case 12 -> {prim2();
                push(bool2scm(x == y));break;}
            case 13 -> {prim2();push(bool2scm((int)x < (int)y));break;}
            case 14 -> {prim2();push((int)x + (int)y);break;}
            case 15 -> {prim2();push((int)x - (int)y);break;}
            case 16 -> {prim2();push((int)x * (int) y);break;}
            case 17 -> {prim2();push((int)x /(int)y);break;}
            case 18 -> {getChar();break;}
            case 19 -> {
                x = ((Object[]) stack)[0];
                putChar((char)(int)x);
                break;}
            case 20 -> {prim1();System.exit((int)x);}
        }
    }
    private static Object printStack(Object stack){
        if (!(stack instanceof Object[])){
            return stack;
        }
        Object x = ((Object[]) stack)[0];
        Object y = ((Object[]) stack)[1];
        Object z =((Object[]) stack)[2];
        return "["+printStack(x)+","+printStack(y)+","+printStack(z)+"]";
    }
    //function that retrieve the byte from input string
    public static int get_byte(String input) {
        pos++;
        return input.charAt(pos);
    }

    public static int get_Code(String input) {
        int byte_value = get_byte(input) - 35;
        if (byte_value < 0) {
            return 57;
        }
        return byte_value;
    }

    public static int get_int(int accumulator, String input) {
        int nextByte = get_Code(input);
        accumulator *= 46;
        accumulator += nextByte;
        if (nextByte < 46) {
            return accumulator;
        }
        return get_int(accumulator - 46, input);
    }

    public static void push(Object value) {
        stack = new Object[]{value,stack,0};
    }

    public static Object pop() {
        Object x = ((Object[])stack)[0];
        Object value = ((Object[])stack)[1];
        stack = value;
        return x;
    }

    public static char putChar(char c) {
        System.out.print(c);
        return c;
    }

    public static void getChar() {
        try {
            int c = System.in.read();
            if (c == -1) {
                push(-1);
            }
            push(c);
        } catch (IOException e) {
            System.out.println("IOexception");
        }
    }
    public static Object f0s(Object y,Object[] x){x[0] = y;return y;}
    public static Object f1s(Object y,Object[] x){x[1] = y;return y;}
    public static Object f2s(Object y,Object[] x){x[2] = y;return y;}

    public static Object bool2scm(boolean x){
        return x ? TRUE : FALSE;
    }
    public static boolean is_rib(Object x){
        return x instanceof Object[];
    }
    public static Object[] scm2list(Object[] l){
        return !Arrays.equals(l, NIL) ? append(new Object[]{l[0]},scm2list(reduceArray(l))) : new Object[]{};
    }
    public static Object[] list_str2scm(String l){
        return !l.isEmpty() ? new Object[]{l.charAt(0),list_str2scm(l.substring(1)),0} : NIL;
    }
    public static String scm2str(Object[] s){
        return chars2str((Object[])s[0]);
    }
    public static String chars2str(Object[] c){
        return Arrays.equals(c,NIL) ? c[0]+chars2str(reduceArray(c)) : "";
    }
    public static Object[] str2scm(String s){
        return new Object[]{char2scm(s),s.length(),3};
    }
    public static Object[] char2scm(String c){
        return !c.isEmpty() ? new Object[]{c.charAt(0),str2scm(c.substring(1)),0} : NIL;
    }
    private static Object[] reduceArray(Object[] array){
        Object[] temp = new Object[array.length-1];
        for(int i = array.length-1; i>1;--i){
            temp[i-1] = array[i];
        }
        return temp;
    }
    private static Object[] append(Object[] a,Object[] b){
        Object[] temp = new Object[a.length+b.length];
        int index = 0;
        for(Object c : a){
            temp[index++] = c;
        }
        for(Object d : b){
            temp[index++] = d;
        }
        return temp;
    }
    private static Object list_tail(int i, Object list){
        if (i==0){
            return list;
        }
        return list_tail(i-1, ((Object[]) list)[1]);
    }
    public static String arrayToString(Object obj) {
        if (obj instanceof Object[]) {
            Object[] arr = (Object[]) obj;
            StringBuilder sb = new StringBuilder("[");
            for (int i = 0; i < arr.length; i++) {
                sb.append(arrayToString(arr[i])); // Recursively process nested arrays
                if (i < arr.length - 1) {
                    sb.append(", ");
                }
            }
            sb.append("]");
            return sb.toString();
        } else {
            return String.valueOf(obj); // Convert non-array objects to strings
        }
    }

}


